types/sorted_cid_set.go

```diff
@@ -5,188 +5,173 @@ import (
+	return len(s.cids)
 }
 
-// Clear removes all entries from the set.
-func (s *SortedCidSet) Clear() {
-	s.s = s.s[:0]
+// ToSlice returns a slice listing the cids in the set.
+func (s TipSetKey) ToSlice() []cid.Cid {
+	out := make([]cid.Cid, len(s.cids))
+	copy(out, s.cids)
+	return out
 }
 
 // Iter returns an iterator that allows the caller to iterate the set in its sort order.
-func (s SortedCidSet) Iter() SortedCidSetIterator {
-	return SortedCidSetIterator{
-		s: s.s,
+func (s TipSetKey) Iter() TipSetKeyIterator {
+	return TipSetKeyIterator{
+		s: s.cids,
 		i: 0,
 	}
 }
 
-// Equals returns true if the set contains the same items as another set.
-func (s SortedCidSet) Equals(s2 SortedCidSet) bool {
-	if s.Len() != s2.Len() {
+// Equals checks whether the set contains exactly the same CIDs as another.
+func (s TipSetKey) Equals(other TipSetKey) bool {
+	if len(s.cids) != len(other.cids) {
 		return false
 	}
-
-	i1 := s.Iter()
-	i2 := s2.Iter()
-
-	for i := 0; i < s.Len(); i++ {
-		if !i1.Value().Equals(i2.Value()) {
+	for i := 0; i < len(s.cids); i++ {
+		if !s.cids[i].Equals(other.cids[i]) {
 			return false
 		}
 	}
-
 	return true
 }
 
-// Contains checks if s2 is a sub-tipset of s
-func (s *SortedCidSet) Contains(s2 *SortedCidSet) bool {
-	for it := s2.Iter(); !it.Complete(); it.Next() {
-		if !s.Has(it.Value()) {
-			return false
+// ContainsAll checks if another set is a subset of this one.
+func (s *TipSetKey) ContainsAll(other TipSetKey) bool {
+	// Since the slices are sorted we can perform one pass over this set, advancing
+	// the other index index whenever the values match.
+	otherIdx := 0
+	for i := 0; i < s.Len() && otherIdx < other.Len(); i++ {
+		if s.cids[i].Equals(other.cids[otherIdx]) {
+			otherIdx++
 		}
 	}
-	return true
+	// otherIdx is advanced the full length only if every element was found in this set.
+	return otherIdx == other.Len()
 }
 
 // String returns a string listing the cids in the set.
-func (s SortedCidSet) String() string {
+func (s TipSetKey) String() string {
 	out := "{"
 	for it := s.Iter(); !it.Complete(); it.Next() {
 		out = fmt.Sprintf("%s %s", out, it.Value().String())
 	}
 	return out + " }"
 }
 
-// ToSlice returns a slice listing the cids in the set.
-func (s SortedCidSet) ToSlice() []cid.Cid {
-	out := make([]cid.Cid, s.Len())
-	var i int
-	for it := s.Iter(); !it.Complete(); it.Next() {
-		out[i] = it.Value()
-		i++
-	}
-	return out
-}
-
 // MarshalJSON serializes the set to JSON.
-func (s SortedCidSet) MarshalJSON() ([]byte, error) {
-	return json.Marshal(s.s)
+func (s TipSetKey) MarshalJSON() ([]byte, error) {
+	return json.Marshal(s.cids)
 }
 
 // UnmarshalJSON parses JSON into the set.
-func (s *SortedCidSet) UnmarshalJSON(b []byte) error {
-	var ts []cid.Cid
-	if err := json.Unmarshal(b, &ts); err != nil {
+func (s *TipSetKey) UnmarshalJSON(b []byte) error {
+	var cids []cid.Cid
+	if err := json.Unmarshal(b, &cids); err != nil {
 		return err
 	}
-	for i := 0; i < len(ts)-1; i++ {
```

Comment:
Blocking: don't we still need this check?  I'm not sure if we're even using JSON serialization anymore but I don't think we're checking the ordering anywhere without this check.

Reply:
Consider deleting these if they are unused.

Reply:
Oops, this was an oversight, thanks. We certainly still use the serialisation, and there are tests for deserialisation, so I fixed it.